File: apps/central/src/models/message.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, sqlx::Type, Default)]
#[sqlx(type_name = "VARCHAR", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
#[allow(dead_code)]
pub enum MessageType {
    #[default]
    Text,
    CallEnded,
    CallMissed,
    CallRejected,
}

impl std::fmt::Display for MessageType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            MessageType::Text => write!(f, "text"),
            MessageType::CallEnded => write!(f, "call_ended"),
            MessageType::CallMissed => write!(f, "call_missed"),
            MessageType::CallRejected => write!(f, "call_rejected"),
        }
    }
}

impl std::str::FromStr for MessageType {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "text" => Ok(MessageType::Text),
            "call_ended" => Ok(MessageType::CallEnded),
            "call_missed" => Ok(MessageType::CallMissed),
            "call_rejected" => Ok(MessageType::CallRejected),
            _ => Err(format!("Unknown message type: {}", s)),
        }
    }
}

#[derive(Debug, Clone, sqlx::FromRow)]
pub struct MessageRow {
    pub id: Uuid,
    pub conversation_id: Uuid,
    pub sender_id: Uuid,
    pub encrypted_content: Vec<u8>,
    pub signature: Vec<u8>,
    pub reply_to_id: Option<Uuid>,
    pub expires_at: Option<DateTime<Utc>>,
    pub sender_chain_id: Option<i32>,
    pub sender_chain_iteration: Option<i32>,
    pub edited_at: Option<DateTime<Utc>>,
    pub message_type: String,
    pub call_id: Option<Uuid>,
    pub call_duration_seconds: Option<i32>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize)]
pub struct Message {
    pub id: Uuid,
    pub conversation_id: Uuid,
    pub sender_id: Uuid,
    pub encrypted_content: Vec<u8>,
    pub signature: Vec<u8>,
    pub reply_to_id: Option<Uuid>,
    pub expires_at: Option<DateTime<Utc>>,
    pub sender_chain_id: Option<i32>,
    pub sender_chain_iteration: Option<i32>,
    pub edited_at: Option<DateTime<Utc>>,
    pub message_type: String,
    pub call_id: Option<Uuid>,
    pub call_duration_seconds: Option<i32>,
    pub reactions: Vec<MessageReaction>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, sqlx::FromRow, Serialize)]
pub struct MessageReaction {
    pub id: Uuid,
    pub message_id: Uuid,
    pub user_id: Uuid,
    pub emoji: String,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, sqlx::FromRow, Serialize)]
pub struct MessageKey {
    pub id: Uuid,
    pub message_id: Uuid,
    pub user_id: Uuid,
    pub encrypted_key: Vec<u8>,
    pub created_at: DateTime<Utc>,
}

File: apps/client/src/features/chat/types/message.ts
export interface ApiMessageReaction {
  id: string;
  message_id: string;
  user_id: string;
  emoji: string;
  created_at: string;
}

export interface Reaction {
  id: string;
  userId: string;
  messageId: string;
  emoji: string;
  createdAt: string;
}

export interface Message {
  id: string;
  conversation_id: string;
  sender_id: string;
  encrypted_content: number[];
  signature: number[];
  reply_to_id: string | null;
  expires_at: string | null;
  sender_chain_id?: number | null;
  sender_chain_iteration?: number | null;
  reactions: ApiMessageReaction[];
  created_at: string;
  encrypted_key?: number[] | null;
  edited_at?: string | null;
  message_type?: string;
  call_id?: string | null;
  call_duration_seconds?: number | null;
  pinned_at?: string | null;
}

export interface MessageKeyData {
  user_id: string;
  encrypted_key: number[];
}

export interface AttachmentMetadata {
  s3_key: string;
  file_size: number;
  encrypted_size: number;
  mime_type: string;
}

export interface SendMessageRequest {
  encrypted_content: number[];
  signature: number[];
  reply_to_id?: string | null;
  expires_at?: string | null;
  sender_chain_id?: number | null;
  sender_chain_iteration?: number | null;
  message_keys?: MessageKeyData[];
  attachment?: AttachmentMetadata;
}

export interface SendMessageResponse {
  id: string;
  created_at: string;
}

export interface EditMessageRequest {
  encrypted_content: number[];
  signature: number[];
  message_keys?: Array<{ user_id: string; encrypted_key: number[] }>;
}

export interface EditMessageResponse {
  edited_at: string;
}

export interface AddReactionRequest {
  emoji: string;
}

export interface ReactionResponse {
  id: string;
  message_id: string;
  user_id: string;
  emoji: string;
  created_at: string;
}

export interface GetMessagesQuery {
  limit?: number;
  before?: string;
}
